datafusion/physical-plan/src/memory.rs

```diff
@@ -236,6 +237,8 @@ impl MemoryExec {
 pub struct MemoryStream {
     /// Vector of record batches
     data: Vec<RecordBatch>,
+    /// Optional memory reservation bound to the data, freed on drop
```

Comment:
I was a little worried at first that this optional API makes it easy to forget to provide reservation. However I see now that the reservation is used only with the recursive CTE case

(not for this PR) In general I wondered if we should *always* have a memory reservation to `MemoryStream`  🤔 I think that would double count batches from a MemTable however, so it isn't an obviously good improvement

Reply:
This concern is quite reasonable; previously there was no need to add a `MemoryReservation` to the `MemoryStream`, and the `with_reservation` function is a bit confused.

I plan to no longer use `MemoryStream` for future recursive CTEs and instead use a new stream class that can read the `WorkTable` multiple times, in order to support 
https://github.com/apache/arrow-datafusion/blob/81b0a011705b17a09f494f550a5382b0c3414597/datafusion/physical-plan/src/recursive_query.rs#L316 

So that the `MemoryStream::with_reservation` API can be removed.